File: retry_with_backoff.rb
# Retry with Exponential Backoff and Jitter
#
# Prevents retry storms by spreading retry attempts over time.
# Without jitter, all failed clients retry at the same instant,
# creating a thundering herd that can collapse an already degraded system.

def retry_with_backoff(max_retries: 5)
  attempt = 0

  begin
    yield
  rescue StandardError => e
    attempt += 1
    raise e if attempt >= max_retries

    # Exponential backoff: 1s, 2s, 4s, 8s...
    base_delay = 2**attempt

    # Jitter: adds random variation of 0-50%
    # This prevents all clients from retrying at the same instant
    jitter = rand * 0.5 * base_delay

    sleep(base_delay + jitter)
    retry
  end
end

# Usage example:
# retry_with_backoff(max_retries: 3) do
#   api_client.fetch_user(user_id)
# end

File: retry_with_backoff.py
# Retry with Exponential Backoff and Jitter
#
# Prevents retry storms by spreading retry attempts over time.
# Without jitter, all failed clients retry at the same instant,
# creating a thundering herd that can collapse an already degraded system.

import random
import time
from typing import Callable, TypeVar

T = TypeVar("T")


def retry_with_backoff(
    fn: Callable[[], T],
    max_retries: int = 5,
) -> T:
    for attempt in range(1, max_retries + 1):
        try:
            return fn()
        except Exception:
            if attempt >= max_retries:
                raise

            # Exponential backoff: 2s, 4s, 8s, 16s...
            base_delay = 2**attempt

            # Jitter: adds random variation of 0-50%
            jitter = random.random() * 0.5 * base_delay

            time.sleep(base_delay + jitter)


# Usage example:
# from retry_with_backoff import retry_with_backoff
#
# def fetch_user():
#     return api_client.get(f"/users/{user_id}")
#
# result = retry_with_backoff(fetch_user, max_retries=3)
